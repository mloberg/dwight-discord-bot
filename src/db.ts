import Keyv from 'keyv';
import { KeyvFile } from 'keyv-file';

import config from './config';
import logger from './logger';

class Database {
    private impl?: Keyv;

    constructor(private readonly uri: string | undefined, private log: typeof logger) {}

    get(key: string, fallback?: unknown) {
        return this.kv.get(key) || fallback;
    }

    set(key: string, value: unknown, ttl?: number) {
        return this.kv.set(key, value, ttl);
    }

    private get kv() {
        if (this.impl) {
            return this.impl;
        }

        this.impl = config.dbUrl?.startsWith('file:')
            ? new Keyv({
                  store: new KeyvFile({
                      filename: this.uri?.substr(5),
                  }),
              })
            : new Keyv(config.dbUrl);

        this.impl.on('error', this.log.error);

        return this.impl;
    }
}

export default new Database(config.dbUrl, logger);
