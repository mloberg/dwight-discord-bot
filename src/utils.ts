export function between(min: number, max: number): number {
    return Math.floor(Math.random()*(max-min+1)+min);
}

export function env(key: string, _default: string): string {
    return process.env[key] || _default;
}

export function rand<T>(items: T[]): T {
    return items[Math.floor(Math.random() * items.length)];
}

export function ucfirst(input: string): string {
    return input.charAt(0).toUpperCase() + input.slice(1).toLowerCase();
}

export function roll(expr: string): number {
    const input = expr.toLowerCase().match(/^(\d+)?d(\d+)([+-×x])?(\d+)?/);
    if (!input) {
        throw new Error("Invalid roll expression");
    }

    const count = Number(input[1] || 1);
    const dice = Number(input[2]);
    const modifier = input[3];
    const mod = Number(input[4]);

    let total = 0;
    for (let index = 0; index < count; index++) {
        total += Math.floor(Math.random() * (dice - 1 + 1)) + 1;
    }

    if ("+" === modifier) {
        total += mod;
    } else if ("-" === modifier) {
        total -= mod;
    } else if ("x" === modifier || "×" === modifier) {
        total = total * mod;
    }

    return total;
}
