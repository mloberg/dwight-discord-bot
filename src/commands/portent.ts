import Command from '../command';
import db from '../db';
import { FriendlyError } from '../error';
import { roll } from '../utils';

export default new Command({
    name: 'portent',
    description: 'Manage portent dice for a Divination Wizard',
    usage: '<show|roll|use> [d20]',
    args: /(?<command>use|roll|show|list) ?(?<roll>\d+)?/,
    examples: ['show', 'roll', 'use 16'],
    async run(message, { groups }) {
        const key = `${message.guild?.name}-${message.author.username}`;
        const results: number[] = groups.command === 'roll' ? [roll('d20'), roll('d20')] : await db.get(key, []);

        if (!results || !results.length) {
            throw new FriendlyError("You don't have any portent rolls. Generate some with `portent roll`.");
        }

        const used = Number(groups.roll);
        if (used) {
            const idx = results.indexOf(used);
            if (idx === -1) {
                throw new FriendlyError(`No portent roll for ${used}. Available: ${results.join(' & ')}`);
            }

            results.splice(idx, 1);
        }

        await db.set(key, results);
        await message.delete();

        return message.reply(results.join(' & '));
    },
});
