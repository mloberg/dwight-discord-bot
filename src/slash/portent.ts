import db from '../db';
import { FriendlyError } from '../error';
import { SlashCommand } from '../types';
import { roll } from '../utils';

const portent: SlashCommand = {
    name: 'portent',
    description: 'Manage portent dice for a Divination Wizard',
    options: [
        {
            name: 'show',
            description: 'Show available portent dice',
            type: 'SUB_COMMAND',
        },
        {
            name: 'roll',
            description: 'Roll new portent dice',
            type: 'SUB_COMMAND',
            options: [
                {
                    name: 'one',
                    description: 'Value of first die',
                    type: 'INTEGER',
                },
                {
                    name: 'two',
                    description: 'Value of second die',
                    type: 'INTEGER',
                },
                {
                    name: 'three',
                    description: 'Value of third die (greater portent)',
                    type: 'INTEGER',
                },
                {
                    name: 'greater',
                    description: 'Roll three dice instead of two',
                    type: 'BOOLEAN',
                },
            ],
        },
        {
            name: 'use',
            description: 'Use a portent die',
            type: 'SUB_COMMAND',
            options: [
                {
                    name: 'dice',
                    description: 'Die value to use',
                    type: 'INTEGER',
                    required: true,
                },
            ],
        },
    ],
    async run(command) {
        const sub = command.options.getSubcommand();
        const key = `${command.guild?.id}-${command.user.id}`;
        const dice: number[] = (await db.get(key)) || [];

        if (sub === 'show') {
            await command.reply(dice.length ? dice.join(', ') : 'No available portent dice.');
            return;
        }

        if (sub === 'roll') {
            const results = [
                command.options.getInteger('one') || roll('d20'),
                command.options.getInteger('two') || roll('d20'),
            ];
            if (command.options.getBoolean('greater')) {
                results.push(command.options.getInteger('three') || roll('d20'));
            }
            await db.set(key, results);
            await command.reply(results.join(', '));
            return;
        }

        const used = command.options.getInteger('dice', true);
        const index = dice.indexOf(used);
        if (index === -1) {
            throw new FriendlyError(`No portent dice for ${used}. Available: ${dice.join(', ')}`);
        }
        dice.splice(index, 1);
        await db.set(key, dice);

        await command.reply(`Remaining: ${dice.join(', ')}`);
    },
};

export default portent;
