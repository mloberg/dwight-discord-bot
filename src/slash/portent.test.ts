import { CommandInteraction } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import db from '../db';
import { FriendlyError } from '../error';
import portent from './portent';

jest.mock('discord.js', () => ({
    CommandInteraction: jest.fn().mockImplementation(() => ({
        guild: {
            id: '1234',
        },
        user: {
            id: '6789',
        },
        reply: jest.fn(),
        options: {
            getSubcommand: jest.fn(),
            getInteger: jest.fn(),
            getBoolean: jest.fn(),
        },
    })),
}));
jest.mock('../db');

const mockDb = mocked(db);

describe('/portent', () => {
    beforeEach(() => {
        mockDb.get.mockReset();
        mockDb.set.mockReset();
    });

    it('is a slash command', () => {
        expect(portent).toMatchSnapshot();
    });

    describe('show', () => {
        it('shows available portent dice', async () => {
            mockDb.get.mockResolvedValue([1, 20]);

            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('show');

            await portent.run(command);
            expect(mockDb.get).toHaveBeenCalledWith('1234-6789');
            expect(command.reply).toHaveBeenCalledWith('1, 20');
        });

        it('shows no available dice', async () => {
            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('show');

            await portent.run(command);
            expect(command.reply).toHaveBeenCalledWith('No available portent dice.');
        });
    });

    describe('roll', () => {
        it('generates portent dice', async () => {
            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('roll');

            await portent.run(command);
            expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/^\d+, \d+$/));
            expect(mockDb.set).toHaveBeenCalledWith('1234-6789', expect.any(Array));
        });

        it('uses given portent dice', async () => {
            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('roll');
            command.options.getInteger.mockReturnValueOnce(1);
            command.options.getInteger.mockReturnValueOnce(2);
            // does not use unless greater portent
            command.options.getInteger.mockReturnValueOnce(3);

            await portent.run(command);
            expect(command.reply).toHaveBeenCalledWith('1, 2');
            expect(mockDb.set).toHaveBeenCalledWith('1234-6789', [1, 2]);
        });

        it('generates greater portent dice', async () => {
            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('roll');
            command.options.getBoolean.mockReturnValue(true);

            await portent.run(command);
            expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/^\d+, \d+, \d+$/));
        });

        it('uses given greater portent dice', async () => {
            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('roll');
            command.options.getBoolean.mockReturnValue(true);
            command.options.getInteger.mockReturnValueOnce(1);
            command.options.getInteger.mockReturnValueOnce(2);
            command.options.getInteger.mockReturnValueOnce(3);

            await portent.run(command);
            expect(command.reply).toHaveBeenCalledWith('1, 2, 3');
            expect(mockDb.set).toHaveBeenCalledWith('1234-6789', [1, 2, 3]);
        });
    });

    describe('use', () => {
        it('removes a portent die', async () => {
            mockDb.get.mockResolvedValue([1, 5, 20]);

            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('use');
            command.options.getInteger.mockReturnValue(5);

            await portent.run(command);
            expect(mockDb.set).toHaveBeenCalledWith('1234-6789', [1, 20]);
            expect(command.reply).toHaveBeenCalledWith('Remaining: 1, 20');
        });

        it('removes one portent die if two match', async () => {
            mockDb.get.mockResolvedValue([20, 20]);

            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('use');
            command.options.getInteger.mockReturnValue(20);

            await portent.run(command);
            expect(mockDb.set).toHaveBeenCalledWith('1234-6789', [20]);
            expect(command.reply).toHaveBeenCalledWith('Remaining: 20');
        });

        it('throws error if no portent die matches', async () => {
            mockDb.get.mockResolvedValue([1]);

            const command = mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('use');
            command.options.getInteger.mockReturnValue(20);

            await expect(portent.run(command)).rejects.toThrowError(
                new FriendlyError('No portent dice for 20. Available: 1'),
            );
        });
    });
});
