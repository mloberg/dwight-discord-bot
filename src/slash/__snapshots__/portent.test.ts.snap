// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/portent is a slash command 1`] = `
Object {
  "description": "Manage portent dice for a Divination Wizard",
  "name": "portent",
  "options": Array [
    Object {
      "description": "Show available portent dice",
      "name": "show",
      "type": "SUB_COMMAND",
    },
    Object {
      "description": "Roll new portent dice",
      "name": "roll",
      "options": Array [
        Object {
          "description": "Value of first die",
          "name": "one",
          "type": "INTEGER",
        },
        Object {
          "description": "Value of second die",
          "name": "two",
          "type": "INTEGER",
        },
        Object {
          "description": "Value of third die (greater portent)",
          "name": "three",
          "type": "INTEGER",
        },
        Object {
          "description": "Roll three dice instead of two",
          "name": "greater",
          "type": "BOOLEAN",
        },
      ],
      "type": "SUB_COMMAND",
    },
    Object {
      "description": "Use a portent die",
      "name": "use",
      "options": Array [
        Object {
          "description": "Die value to use",
          "name": "dice",
          "required": true,
          "type": "INTEGER",
        },
      ],
      "type": "SUB_COMMAND",
    },
  ],
  "run": [Function],
}
`;
